### PR TITLE
Data Collectors : Load data collectors using ExtensionsPath

### PR DESCRIPTION
**Pull request by harshjain2**
1. Load data collectors using ExtensionsPath that will come from RunSettings->RunConfiguration->ExtensionsPath.

2. Load all the data collectors that follow *Collector.dll naming convention and create a cache.

Related to : https://github.com/Microsoft/vstest/issues/309

### CODE REVIEW

**Comment by harshjain2** [comment]
@dotnet-bot test this please

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 1]
```diff
@@ -371,7 +401,7 @@ private void LoadAndInitialize(DataCollectorSettings dataCollectorSettings)
             {
                 if (EqtTrace.IsErrorEnabled)
                 {
-                    EqtTrace.Error("DataCollectionManager.LoadAndInitialize: exception while creating data collector {0} : {1}", collectorTypeName, ex);
+                    EqtTrace.Error("DataCollectionManager.LoadAndInitialize: exception while creating data collector {0} : {1}", collectorTypeName, ex.Message);
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: Shouldn't we need stacktrace?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 2]
```diff
@@ -628,5 +658,40 @@ private void LogAttachments(List<AttachmentSet> attachmentSets)
                 }
             }
         }
+
+        /// <summary>
+        /// Initializes the Data Collector Cache.
+        /// </summary>
+        /// <param name="extensionsPath">Extensions path.</param>
+        private void InitializeDataCollectorCache(string extensionsPath)
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: This doesn't follow the existing extensibility convention of `TestPluginManager` etc.. Is it intentional?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 3]
```diff
@@ -628,5 +658,40 @@ private void LogAttachments(List<AttachmentSet> attachmentSets)
...
+        {
+            var extpaths = string.IsNullOrWhiteSpace(extensionsPath) ? new List<string>() : extensionsPath.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries).ToList();
+
+            // add current directory and extensions folder path. 
+            // todo : remove this once extensions path starts coming from vstest.console.
+            var currentDir = Path.GetDirectoryName(typeof(DataCollectionManager).GetTypeInfo().Assembly.Location);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: Create an issue for this please? We need to bring the support for `ExtensionPath` for all other extensibility types and remove the hardcoding for `Extensions` directory. /cc: @Faizan2304

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 4]
```diff
@@ -628,5 +658,40 @@ private void LogAttachments(List<AttachmentSet> attachmentSets)
...
+            var extDir = Path.Combine(currentDir, "Extensions");
+
+            extpaths.Add(currentDir);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: Why is this required?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 5]
```diff
@@ -628,5 +658,40 @@ private void LogAttachments(List<AttachmentSet> attachmentSets)
...
+                {
+                    var extensions = this.fileHelper.EnumerateFiles(extPath, TestPlatformConstants.DataCollectorRegexPattern, SearchOption.TopDirectoryOnly);
+
+                    foreach (var extension in extensions)
+                    {
+                        var dataCollectors = this.dataCollectorLoader.FindDataCollectors(extension);
+
+                        foreach (var datacollector in dataCollectors)
+                        {
+                            this.dataCollectorCache[datacollector.Item1] = datacollector.Item2;
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: Do we need to log the extensions that match? And the directories we look into?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 6]
```diff
@@ -517,7 +547,7 @@ private void SendEvent(DataCollectionEventArgs args)
 
                     if (EqtTrace.IsErrorEnabled)
                     {
-                        EqtTrace.Error("DataCollectionManager.GetEnvironmentVariables: Failed to get variable for Collector '{0}': {1}", dataCollectorType, ex);
+                        EqtTrace.Error("DataCollectionManager.GetEnvironmentVariables: Failed to get variable for Collector '{0}': {1}", dataCollectorType, ex.Message);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: Why is this change required?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorLoader.cs [hunk 7]
```diff
@@ -29,18 +57,102 @@ public DataCollector Load(string location, string assemblyQualifiedName)
 #else
                 assembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(location);
 #endif
-                var dataCollectorType = assembly.GetTypes().Where(t => t.AssemblyQualifiedName.Equals(assemblyQualifiedName)).FirstOrDefault();
-                dataCollectorInstance = Activator.CreateInstance(dataCollectorType) as DataCollector;
+
+                foreach (var type in assembly.GetTypes())
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorLoader.cs]: This is not following the convention used by other extension loaders. Is that intentional?
